staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,299 @@
+			// if there is an error while setting up or handling the watches, this will ensure that we will process the config file.
+			defer d.queue.Add(workqueueKey)
+			klog.ErrorS(err, "Failed to watch encryption config file, will retry later")
+		}
+	}, time.Second, ctx.Done())
+
+	<-ctx.Done()
+}
+
+func (d *DynamicKMSEncryptionConfigContent) watchEncryptionConfigFile(stopCh <-chan struct{}) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	if err = watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
+
+	for {
+		select {
+		case event := <-watcher.Events:
+			if err := d.handleWatchEvent(event, watcher); err != nil {
+				return err
+			}
+		case err := <-watcher.Errors:
+			return fmt.Errorf("received fsnotify error: %v", err)
+		case <-stopCh:
+			return nil
+		}
+	}
+}
+
+func (d *DynamicKMSEncryptionConfigContent) handleWatchEvent(event fsnotify.Event, watcher *fsnotify.Watcher) error {
+	// This should be executed after restarting the watch (if applicable) to ensure no file event will be missing.
+	defer d.queue.Add(workqueueKey)
+
+	// return if file has not been removed or renamed.
+	if event.Op&(fsnotify.Remove|fsnotify.Rename) == 0 {
+		return nil
+	}
+
+	if err := watcher.Remove(d.filePath); err != nil {
+		klog.V(2).InfoS("Failed to remove file watch, it may have been deleted", "file", d.filePath, "err", err)
+	}
+	if err := watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
+
+	return nil
+}
+
+// runWorker to process file content
+func (d *DynamicKMSEncryptionConfigContent) runWorker() {
+	for d.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes file content when there is a message in the queue.
+func (d *DynamicKMSEncryptionConfigContent) processNextWorkItem() bool {
+	// key here is dummy item in the queue to trigger file content processing.
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	// get context to close the old transformers.
+	ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
+	transformerOverrides, kmsPluginHealthzChecks, processedConfigHash, kmsMaxTimeout, totalKMSPlugins, err := d.processEncryptionConfig(ctx)
+	if err != nil {
+		// close transformers to avoid leaking resources.
+		closeTransformers()
+
+		utilruntime.HandleError(fmt.Errorf("error processing encryption config file %s: %v", d.filePath, err))
+		// add dummy item back to the queue to trigger file content processing.
+		d.queue.AddRateLimited(key)
+
+		// return true to continue processing the queue.
+		return true
+	}
+
+	// if transformerOverrides is nil, it means that the encryption config file has not changed.
+	if transformerOverrides == nil {
+		d.queue.Forget(key)
+		return true
+	}
+
+	err = d.updateTransformers(
```

Comment:
the nesting of logic and calling closeTransformers() at uneven levels in error cases (processNextWorkItem → updateTransformers → healthz) is pretty hard to follow

it would be easier to be confident we've handled unchanged and error cases if this was called in a single place in processNextWorkItem, and processNextWorkItem was responsible for the checks and cleanup in error cases

```go
ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
var (
  updatedEffectiveConfig bool
  err error
)
defer func(){
  if updatedEffectiveConfig {
    // TODO: increment success metric
  } else {
    // avoid leaking on error or unchanged
    closeTransformers()
  }

  if err != nil {
    // TODO: increment failure metric
  }
}()

..., configChanged, err = processEncryptionConfig(...)
if err != nil {
  return
}
if !configChanged {
  return
}
err = healthz(...)
if err != nil {
  return
}
d.dynamicTransformers.Set(...)
d.encryptionConfigHash = processedConfigHash
updatedEffectiveConfig = true
return
```

Reply:
I had it consolidated before but then separated it to make healthz testable. But this is better. I updated it. Thanks!